File: dp/322.py
class Solution:
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        
        #初始化时, 使数组中预估数量都比最终数额大一, 也就是一个超过可能情况最大值的一个数量, 这样可以再无法组合的情况时用于判断返回-1
        ret = [amount+1 for i in range(0, amount+1)]
        for i in range(0, amount+1):
            if i == 0:
                ret[i] = 0
            else:
                for coin in coins:
                    if i >= coin:
                        ret[i] = min(ret[i-coin] + 1, ret[i])

        if ret[amount] == amount+1:
            return -1
        return ret[amount]        

File: dp/322.cc
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int dp[amount+1];
        dp[0] = 0;
        for (int i = 1; i <= amount; ++i) {
            dp[i] = INT_MAX;
            for (const auto &coin :coins) {
                if ((i-coin) < 0) {
                    continue;
                }
                if (dp[i-coin] != INT_MAX) {
                    dp[i] = std::min(dp[i], dp[i-coin]+1);    
                }
            }
        }
        if (INT_MAX == dp[amount]) {
            return -1;
        }
        return dp[amount];
    }
};
